### PR TITLE
Add explicit multiply-reduce GEMM kernel

### PR DESCRIPTION
**Pull request by brunomazzottiamd**
Add kernel that implements GEMM with explicit multiply-reduce instructions for small block sizes. Such
small block sizes aren't natively supported by `tl.dot` operator.

Despite being numerically correct, this kernel performed worse than a corresponding GEMM kernel that
used `tl.dot` with minimum block size equal to $16$:

**MI300 Results for FP16:**

| trans   |   M |     N |     K |         Dot GiBps |   Multiply-Reduce GiBps |   Speedup |
|:--------|----:|------:|------:|------------------:|------------------------:|----------:|
| TN      |   1 |  8192 | 28672 |           3491.33 |                 2869.87 |      0.82 |
| TN      |   1 |  6144 |  6144 |           3858.22 |                 2673.33 |      0.69 |
| TN      |   1 |  4096 |  4096 |           2352.54 |                 1680.93 |      0.71 |
| TN      |   2 | 16384 | 16384 |           3412.17 |                 3318.44 |      0.97 |

The code is based on **tune_gemm** `matmul_kernel` from commit `cf44637` (see `triton-mlir` branch).

### CODE REVIEW

File: python/perf-kernels/multreduce_matmul_kernel.py
Comment (vgokhale) [inline python/perf-kernels/multreduce_matmul_kernel.py:53]: Would this be useful for BLOCK_N or K < 16?

File: python/perf-kernels/multreduce_matmul_kernel.py
Comment (vgokhale) [inline python/perf-kernels/multreduce_matmul_kernel.py:26]: Can you remove the other GEMM code like splitK and code for large M/N?

File: python/perf-kernels/multreduce_matmul_kernel.py
Comment (brunomazzottiamd) [reply python/perf-kernels/multreduce_matmul_kernel.py:53]: In terms of performance I really don't know because I didn't test it for `BLOCK_SIZE_N < 16` or `BLOCK_SIZE_K < 16`. At least it'll make the kernel compile correctly because `tl.dot` won't compile for such sizes.

Please check these code snippets and you'll fully understand:
* https://github.com/triton-lang/triton/blob/86f08228102a29b43e998a457981615753f5ff2f/python/triton/language/semantic.py#L1375
* https://github.com/triton-lang/triton/blob/main/third_party/amd/backend/compiler.py#L14

File: python/perf-kernels/multreduce_matmul_kernel.py
Comment (brunomazzottiamd) [reply python/perf-kernels/multreduce_matmul_kernel.py:26]: Of course, I can remove these pieces. Please let me remove them in another development branch and check correctness again. After that, I'm going to update this PR.

Comment (brunomazzottiamd) [reply python/perf-kernels/multreduce_matmul_kernel.py]: Now it's done. `SPLIT_K` was removed by 9cbebf09ce9a11b31890dd21a4a1e15babbb28ca and `GROUP_SIZE_M` was removed by dd161e502a3bbc6cabfae9ded4d006bdadd6973c.

File: python/perf-kernels/README.md
Comment (vgokhale) [inline python/perf-kernels/README.md:73]: I think we can remove references to this for now. This operator won't be used if dot works so comparing it doesn't make sense. Thye rest of the readme is fine.

File: python/perf-kernels/multreduce_matmul_kernel.py
Comment (vgokhale) [inline python/perf-kernels/multreduce_matmul_kernel.py:46]: You can also remove this. We don't want to use dot here at all.

File: python/perf-kernels/README.md
Comment (brunomazzottiamd) [reply python/perf-kernels/README.md]: Done by dc7a5f7e5.

File: python/perf-kernels/multreduce_matmul_kernel.py
Comment (brunomazzottiamd) [reply python/perf-kernels/multreduce_matmul_kernel.py]: Done by 2bba0d0b5.

Comment (brunomazzottiamd) [reply python/perf-kernels/multreduce_matmul_kernel.py]: @vgokhale, even after all the work I still don't understand the purpose of this kernel we created. Could you please share your thoughts about this?

The facts are:
* `tl.dot` doesn't work for small block sizes.
* We have a kernel that computes the required dot product without relying on `tl.dot`, i.e. relying on explicit elementwise multiplication and sum reduction operations.
* In terms of performance, the proposed kernel is worse than using `tl.dot` with the minimum block size.

What are the advantages of the proposed kernel? If I were a Triton user on AMD hardware, I'd use `tl.dot` with the minimum block size instead of explicit multiply-reduce approach.

**Comment by brunomazzottiamd** [comment]
The related issue is https://github.com/ROCm/triton-internal/issues/169.

**Review by vgokhale — APPROVED** [review]

File: python/perf-kernels/multreduce_matmul_kernel.py
Comment (brunomazzottiamd) [reply python/perf-kernels/multreduce_matmul_kernel.py]: Vinayak answered this question to me in a 1:1 meeting.

The proposed kernel will be a reference for further compiler developments and can also serve as a baseline for TorchInductor improvements. Furthermore, we can analyze its performance again in the future as the compiler gets better over time.

**Comment by brunomazzottiamd** [comment]
@xiaohuguo2023 told me we can merge PRs by ourselves once we have one approval. I'll do it.

**Pull request merged**
